Add exponential backoff when retrieving a ICS to catch errors

Hello
On my side i use a calendar from my Google drive account updated hourly.

Sometimes the fetchSourceCalendars in : UrlFetchApp.fetch(url).getContentText(); get error 403.
meaning the calendar is then empty. As a result all event are droped from the target calendar. :-(
 To prevent it is see below changes in the fetchSourceCalendars.
and testing the result in code.gs exiting the program if calendar cannot be loaded (Note : i'm using only one calendar ) 
```
 //------------------------ Fetch URL items ------------------------
    var responses = fetchSourceCalendars(sourceCalendarURLs);
    if (responses.length == null)
    { 
      Logger.log ('Issue on geting Cal from Drive');
      Logger.log ('Exit Main');
      return false;
    }
```
As result the calendar retrieval is more robust...

```
function fetchSourceCalendars(sourceCalendarURLs){
  var vsuccess = false;
  var vcount = 0;
  var MAX_TRIES = 10;
  var result = [];
  for each (var url in sourceCalendarURLs){
    url = url.replace("webcal://", "https://");      


  while(!vsuccess && vcount++ < MAX_TRIES){
     try{ 
			var urlResponse = UrlFetchApp.fetch(url).getContentText();
            Logger.log("loading: " + url);
			//------------------------ Error checking ------------------------
			if(!urlResponse.includes("BEGIN:VCALENDAR")){
				Logger.log("[ERROR] Incorrect ics/ical URL: " + url);
			}
			else{
				result.push(urlResponse);
				vsuccess = true;
			}
     } catch (e){
       Logger.log("Fetch Cal retry");
       Utilities.sleep(1000);
     }
  }
  if(!vsuccess){
    Logger.log("Still failed after 10 retry");
    return false;
  }
```

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -49,22 +49,28 @@ function deleteAllTriggers(){
 function fetchSourceCalendars(sourceCalendarURLs){
   var result = []
   for (var url of sourceCalendarURLs){
-    url = url.replace("webcal://", "https://");      
+    url = url.replace("webcal://", "https://");
     
-    try{
-      var urlResponse = UrlFetchApp.fetch(url).getContentText();
-      //------------------------ Error checking ------------------------
-      if(!urlResponse.includes("BEGIN:VCALENDAR")){
-        Logger.log("[ERROR] Incorrect ics/ical URL: " + url);
+    callWithBackoff(function() {
+      var urlResponse = UrlFetchApp.fetch(url);
+      if (urlResponse.getResponseCode() == 200){
+        var urlContent = urlResponse.getContentText();
+        if(!urlContent.includes("BEGIN:VCALENDAR")){
+          Logger.log("[ERROR] Incorrect ics/ical URL: " + url);
+          return;
+        }
+        else{
+          result.push(urlContent);
+          Logger.log("Result: " + result.length);
+          return;
+        }     
       }
-      else{
-        result.push(urlResponse);
+      else{ //Throw here to make callWithBackoff run again
+        throw "Error: Encountered " + urlReponse.getReponseCode() + " when accessing " + url; 
       }
-    }
-    catch(e){
-      Logger.log(e);
-    }
+    }, 2);
   }
+  
   return result;
 }
 

